### ISSUE TITLE
[Accessibility Audit] Code Intel: Going to a definition from a reference

### ISSUE THREAD

**Issue by umpox** [sourcegraph/sourcegraph#33509]
### Steps to replicate journey

Taken from the [user journey list document](https://docs.google.com/document/d/1kA6aVOAgID_uPm-d6uEC1DG6WHjJ8fPT0UAXQcML_KQ/edit?usp=sharing). Use this for further context.

- Hover over an identifier.
- Click on the Go to Definition button in the Code Intel Popover. [Optional: If you hover over the identifier in the definition, the button will be inactivated and the label will be “You are at the definition”]

### How to audit

Follow the instructions here: [Auditing a user journey](https://docs.sourcegraph.com/dev/background-information/web/accessibility/how-to-audit#auditing-a-user-journey)

**Note:** We are planning on arranging thorough third-party audit in the future, so our focus here is just to find the _majority_ of accessibility issues. Don't worry if you feel you haven't found 100% of potential issues, it is better to focus on the core essentials to complete the journey rather than spending lots of time going through every possible problem.

### How to raise discovered problems

Follow the instructions here: [Raising an accessibility bug](https://docs.sourcegraph.com/dev/background-information/web/accessibility/how-to-audit#raising-a-bug)

### Any other questions?

See the [audit FAQ](https://docs.google.com/document/d/1nCi7TvYRpj9eG0_QtgqK1c4L28EiJKVuxyuvg1lAGgo/edit#) - add your question if it isn't there already!

**Comment by sourcegraph-bot-2** [comment]
Heads up @macraig - the "team/code-intelligence" label was applied to this issue.

**Comment by efritz** [comment]
@varungandhi-src drove an initial audit and I took notes:

- [ ] On the blob view we have no headers other than the sg logo, so it's very difficult to navigate to the code. We suggest to add headers to the following: products layer (top bar), the search bar layer, the navigation and metadata layer (the filepath breadcrumbs + brain dot, github link, etc), the sidebar, the code view, and the reference panel.
- [ ] The code is formatted as a table, which might be correct. It seems like we can navigate down the correct rows without having to read line numbers or gutters each time. Not sure if there was something else notable to mention here (will defer to @varungandhi-src's opinions on this as he was driving).
- [ ] Literally everything reads a clickable but punctuation does nothing
- [ ] It reads "quotation mark", "context", "quotation mark" which seems really messy. Strangely, it reads "fmt" as `F.M.T.` in `./cmd/frontend/backend/repos.go` in sg/sg repo. EDIT: We've found that triggering the popover may refine the highlighted range, which causes the quotation marks to be rendered differently. Specifically, it reads out the name of the import as expected, then breaks it into three tokens once the document highlights have been rendered.
- [ ] When the code intel popover appears there is no indication from the screen reader. We should indicate that something has appeared, as well as what buttons are available. Related: when opening the popover the entire page is focused (previous thought: going from the popover _back_ to the code selects the entire blob view, not the link that triggered the popover).
- [ ] Unclear how to dismiss popover once it's visible, and it doesn't seem re-triggerable if it's already open
- [ ] When you click on go to def it opens a new tab, it announces "you are in <package> in side of web content". We should instead announce the file and/or the repository you were navigated to. Same with back button (possibly)
- [ ] Hovering over punctuation ')' will say there are "two actions". The first one doesn't do anything, the second one is a right click action.
- [ ] "You are at the definition" says that it's clickable but it's disabled. This button seems like it's skipped with tab navigation (but not arrow navigation). Should it be present in tab navigation as well?
- [ ] The header of the popover (the signature) says it's clickable but does nothing. So is the hover text. Why is everything clickable?!
- [ ] The signature `struct { RedirectURL string }` breaks the last brace into a different element, weirdly.
- [ ] If you click "Find Refs" it doesn't focus the reference panel and the focus appears to be on a random usage within the code view. Reference panel opening is not announced whether it's open already or not.
- [ ] Mix search and precise changes state but does not announce any change in the state. Check if we can add aria-live or something.
- [ ] `Sourcegraph logo, sourcegraph clickable image` could have a better aria label
- [ ] On ref panel close button, it gives you options to exit the "web area" - if you go up to the parent you can't re-enter the ref panel. Is this expected?
- [ ] You also can't enter web content from the ref panel header (you get audible bell feedback that there's nothing to do)
- [ ] In ref panel, `Type to filter by filename` announces it's a text box, but would be nice to say how many files there are to filter, and how many matches there are as we type.
- [ ] Collapsible reads out strangely, as `Close <header>` instead of `<header>, click to collapse this content`, which would be much better. Additionally (on us) the text `(n displayed)` should probably be `(n matches)` or `(n occurrences)` to be more clear.
- [ ] Displaying per-repo match counts might be nice as well (we have them per-file and per type (def/ref/impl))
- [ ] Navigating down to a match will say "list item" but not read out the line number or the text
- [ ] Clicking on a match in the ref panel opens content to the right but does not announce a change
- [ ] Tab hierarchy is not changed when the file content pane opens on the right side of the ref panel (we should tab from a line to its match, not down to other matches; @varungandhi-src mentioned this as an opinion)
- [ ] Maybe update text to be "References (5 occurrences *over 3 repositories*)". We should also probably indicate whether there is a load more option here
- [ ] Collapsible (and some other text elements it seems) allows right click
- [ ] ~Tab breaks when closing a collapsible~ Could not reproduce again

**Comment by umpox** [comment]
@varungandhi-src @efritz Did you find any way to navigate directly to a reference using only a keyboard? Curious if is possible in the current state. My understanding is that we lazily check for references on hover, so we wouldn't be able to make these automatically keyboard navigable on page load.

This is a good example of how useful this would be: https://source.chromium.org/chromium/chromium/src/+/main:apps/app_lifetime_monitor_factory.h;drc=a341e189c99581b8f48a8b66200a5140451a3622;l=22. It is really fast to just <kbd>tab</kbd> through each reference.

Maybe there's a simpler, basic, approach we could take to make our code blobs more navigable for now? E.g. leverage our code highlighting so we can tab through variables?

**Comment by varungandhi-src** [comment]
To make sure I'm understanding your point right, the question is purely for tab-based navigation, not for screen-reader stuff? (Screen reader users can use Ctrl + Option + Arrow keys to navigate the table). We don't support that right now. I'll add that as a sub-issue.

> My understanding is that we lazily check for references on hover, so we wouldn't be able to make these automatically keyboard navigable on page load.

I briefly tested this and your point is correct; we're not issuing a GraphQL Ranges query until we get a hover.

> This is a good example of how useful this would be: https://source.chromium.org/chromium/chromium/src/+/main:apps/app_lifetime_monitor_factory.h;drc=a341e189c99581b8f48a8b66200a5140451a3622;l=22. It is really fast to just tab through each reference.

Could you attach a screen recording of you doing this? I tried navigating using tab and shift-tab, but was unable to cycle through references like you mention.

**Comment by umpox** [comment]
> Could you attach a screen recording of you doing this? I tried navigating using tab and shift-tab, but was unable to cycle through references like you mention.

Sorry, maybe references is the wrong term, tokens I guess? Here's a recording, notice how I can just use the keyboard to navigate between each item.

https://user-images.githubusercontent.com/9516420/168242048-74aa974a-908a-4e07-8bb6-f0afdadbc69e.mp4

As far as I can tell, it's not possible to trigger the reference panel using _only_ a keyboard?

I think we fundamental problem we have is that:
1. All code is just a `<span>` be default
2. On hover, if that code has hover content, i.e. we could show the intel popover, it effectively will act as a link - we [update the cursor and it will navigate on click](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/web/src/components/WebHoverOverlay/WebHoverOverlay.tsx?L84%3A1-94%3A1=&utm_source=VSCode-2.2.2).

I wonder if there's any scenario where we can be confident that (barring any errors) a token will **always** trigger the intel popover when hovered (e.g. every variable?). If we knew that, then we could make these links before the user hovers them, and that'd mean they could also tab through them like in that video above.

**Comment by varungandhi-src** [comment]
Was just getting around to filing the issues I didn't file earlier, and I tried to reproduce your Tab-based navigation for Google code search. Looks like it works in Chrome but not in Firefox, that's why I wasn't able to repro it earlier. But yeah, I've filed an issue for it.
- https://github.com/sourcegraph/sourcegraph/issues/36587

I'm afraid I don't know what the right/best solution is on how to implement it. 😓

**Comment by varungandhi-src** [comment]
Closed this since I think all the issues related to Go to Definition have been filed. I will continue filing more issues as part of the other remaining code intel journeys, but none of them seem specific to Go to Definition.
